modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientPartitionAwarenessDiscoveryTest.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.client.thin;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.ignite.client.ClientAddressFinder;
+import org.apache.ignite.client.ClientCache;
+import org.apache.ignite.configuration.ClientConfiguration;
+import org.apache.ignite.internal.IgnitionEx;
+import org.apache.ignite.internal.processors.cache.IgniteInternalCache;
+import org.junit.Test;
+
+import static org.apache.ignite.internal.processors.odbc.ClientListenerProcessor.CLIENT_LISTENER_PORT;
+
+/**
+ * Test partition awareness of thin client on changed topology.
+ */
+public class ThinClientPartitionAwarenessDiscoveryTest extends ThinClientAbstractPartitionAwarenessTest {
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * Test that client use channels to all running nodes while new nodes start
+     */
+    @Test
+    public void testClientDiscoveryNodesJoin() throws Exception {
+        for (int i = 0; i < MAX_CLUSTER_SIZE; ++i) {
+            startGrid(i);
+            awaitPartitionMapExchange();
+
+            int[] workChannels = IntStream.rangeClosed(0, i).toArray();
+
+            if (i == 0)
+                initClient(getClientConfigurationWithDiscovery(), workChannels);
+            else {
+                detectTopologyChange();
+                awaitChannelsInit(workChannels);
+            }
+
+            testPartitionAwareness(workChannels);
+        }
+    }
+
+    /**
+     * Test that client use channels to all running nodes while nodes stop
+     */
+    @Test
+    public void testClientDiscoveryNodesLeave() throws Exception {
```

Comment:
We don't check discovery by this test at all. If we replace `getClientConfigurationWithDiscovery()` with `getClientConfiguration(0, 1, 2, 3)` result will be the same, test will pass. To check discovery I think we should exclude additional nodes by discovery (but not stop it) on topology change and check that this node is excluded from requests too.

Reply:
Yes, it's ok that test will pass with default configuration too. It's like regression test, that discovery work expectedly. The case you describe will work the same way, it will pass both for default and discovery configurations, as both of them will configure the same set of channels. But it is a good test case and I've added it too.